docs: add release notes for 0.21.0 (#6365)
determined-ai/determined: docs/release-notes.rst

 Release Notes
###############

**************
 Version 0.21
**************

Version 0.21.0
==============

**Release Date:** March 27, 2023

**Breaking Changes**

-  Cluster: K80 GPUs are no longer supported.

> -  Remove all old PATCH endpoints under `/agents*`. This covers the PATCH APIs for enabling and
>    disabling slots. Users should use the new APIs under `/api/v1/agents`

Reviewer: ```suggestion
-  API: Remove all old PATCH endpoints under ``/agents*``, including the APIs for enabling and
   disabling slots. Users should use the new APIs under ``/api/v1/agents``.
```

> -  Trial API: ``on_validation_step_start`` and ``on_validation_step_end`` callbacks on

Reviewer: ```suggestion
-  API: The ``on_validation_step_start`` and ``on_validation_step_end`` callbacks on
```

   ``PyTorchTrial`` and ``DeepspeedTrial`` were deprecated in 0.12.12 (Jul 2020) and have been
   removed. Please use ``on_validation_start`` and ``on_validation_end`` instead.

> -  ``records_per_epoch`` has been dropped from PyTorch code paths. We were previously using this

Reviewer: ```suggestion
-  Trial API: ``records_per_epoch`` has been dropped from PyTorch code paths. We were previously using this
```

   value internally to estimate epoch lengths. We are now using the chief worker's epoch length as
   the epoch length.

-  ``average_training_metrics`` is no longer configurable. This value previously defaulted to false
   and was dropped to simplify the training API. We always average training metrics now.

> -  Trial API: the unused ``latest_training`` field has been removed from ``GetTrial`` and

Reviewer: ```suggestion
-  API: The unused ``latest_training`` field has been removed from the ``GetTrial`` and
```

   ``GetExperimentTrials`` APIs due to slow performance.

**Bug Fixes**

> -  CLI: Fix an issue where ``det user change-password`` will return an authentication error when
>    trying to change current user's password.

Reviewer: ```suggestion
-  CLI: Fix an issue where ``det user change-password`` would return an authentication error when
   trying to change the current user's password.
```

**Improvements**

> -  CLI: Default to provisioning Nvidia T4 GPU instances.
> 
>    -  Command line deployments will now default to provisioning Nvidia T4 GPU instances instead of

Reviewer: ```suggestion
-  CLI: Command-line deployments will now default to provisioning Nvidia T4 GPU instances instead of
```
and adjust indentation of the following lines.

      K80 instances. This change is intended to improve the performance/cost and driver support of
      the default deployment.

-  Kubernetes: Ease permission requirements in Kubernetes so master no longer requires access to all
   Kubernetes namespaces. This only affects custom modified Helm chart configurations.

> -  Checkpoints: Improved performance of checkpoint insertion and deletion.

Reviewer: ```suggestion
-  Checkpoints: Improve performance of checkpoint insertion and deletion.
```

**New Feature**

> -  PyTorchTrial and DeepSpeedTrial APIs: Deprecate ``TorchWriter`` and add a PyTorch

Reviewer: ```suggestion
-  API: Deprecate ``TorchWriter`` and add a PyTorch
```

   ``SummaryWriter`` object to ``PyTorchTrialContext`` and ``DeepSpeedTrialContext`` that we manage
   on behalf of users. See :func:`~determined.pytorch.PyTorchTrialContext.get_tensorboard_writer`
   for details.

> -  Introduces :class:`~determined.pytorch.Trainer`, a high-level training API for ``PyTorchTrial``

Reviewer: ```suggestion
-  API: Introduce :class:`~determined.pytorch.Trainer`, a high-level training API for ``PyTorchTrial``
```

>    that allows for python-side training loop customizations and includes support for off-cluster

Reviewer: ```suggestion
   that allows for Python-side training loop customizations and includes support for off-cluster
```

   local training.

**Removed Features**

-  The following methods of :class:`~determined.experimental.client.Checkpoint`,
   :class:`~determined.experimental.client.Model`, and
   :class:`~determined.experimental.client.ModelVersion` were deprecated in 0.17.9 (Feb 2022) and
   are now removed:

      -  ``Checkpoint.load()``
      -  ``Checkpoint.load_from_path()``
      -  ``Checkpoint.parse_metadata()``
      -  ``Checkpoint.get_type()``
      -  ``Checkpoint.from_json()``
      -  ``Model.from_json()``

>       -  ``ModelVersion.from_json()``

Reviewer: Reduce indentation by 3.
